client/gtk/src/main.rs

```diff
@@ -44,32 +44,75 @@ impl Win {
     fn handle_action(&mut self, action: Action) -> () {
         match action {
             Action::AddMessage(msg) => {
-                let text_buffer = self.widgets.messages.get_buffer().unwrap();
-                let message = format!("{}: {}\n", msg.author, msg.content);
-                text_buffer.insert(&mut text_buffer.get_end_iter(), &message);
+                self.push_message(&msg.author, &msg.content);
+            }
+            Action::AddRoom(room) => {
+                self.push_message("Info", &format!("Joined room {}", room.0));
+
+                self.model.room = Some(room);
+                let txt: &str = &format!("#{}", room.0);
+                let room_label = Label::new(Some(txt));
+                self.widgets.rooms.insert(&room_label, -1);
+                self.model.room_list.push(room);
+                room_label.show_all();
             }
             Action::Error(_error) => {}
         }
     }
+
+    fn push_message(&self, author: &str, content: &str) {
+        let message = self.build_message(author, content);
+
+        let separator = Separator::new(Orientation::Horizontal);
+
+        separator.show_all();
+        message.show_all();
+
+        self.widgets.messages.insert(&separator, -1);
+        self.widgets.messages.insert(&message, -1);
+    }
+
+    fn build_message(&self, author: &str, content: &str) -> Grid {
+        let grid = Grid::new();
+        grid.insert_column(0);
+        grid.insert_column(1);
+        grid.insert_column(2);
```

Comment:
Are there three columns? Don't we just need two currently (author and content)?

Reply:
Right yeah. Not sure why I added the third :P